src/Http/Authentication.Core/src/AuthenticationService.cs

```diff
@@ -77,8 +80,26 @@ public virtual async Task<AuthenticateResult> AuthenticateAsync(HttpContext cont
             var result = await handler.AuthenticateAsync();
             if (result != null && result.Succeeded)
             {
-                var transformed = await Transform.TransformAsync(result.Principal);
-                return AuthenticateResult.Success(new AuthenticationTicket(transformed, result.Properties, result.Ticket.AuthenticationScheme));
+                var principal = result.Principal;
+                var doTransform = true;
+                if (Options.ApplyClaimsTransformationOnce)
+                {
+                    _transformCache = _transformCache ?? new HashSet<ClaimsPrincipal>();
```

Comment:
Maybe this isn't worth doing at all, seems like hashsets are lazy init'd to be very small already, so we can probably just always new it up, since the overhead if we don't use it is low from this comment:

https://github.com/dotnet/corefx/blob/2d6fe6b5cc3adba461bb8737c7e5c947992d4c7a/src/System.Collections/src/System/Collections/Generic/HashSet.cs#L21

Reply:
So the places where you might save allocating is when the service is allocated but never called, or called, but the request is not authenticated. Anonymous is common, so might as well do `_transformCache ?= new HashSet<ClaimsPrincipal>();`

Reply:
Sure